### PR TITLE
New consensus timeout bugfixes

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Address a few outstanding comments with respect to mainly timeout message handling. 

https://github.com/Concordium/concordium-node/pull/751#discussion_r1210363933
https://github.com/Concordium/concordium-node/pull/751#discussion_r1210380258
https://github.com/Concordium/concordium-node/pull/751#discussion_r1211117324
https://github.com/Concordium/concordium-node/pull/751#discussion_r1211131183

## Changes

Corrected the code according to the above mentioned comments. 

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by chportma — APPROVED** [review]
It looks good! There is just one comment in the code which is now inaccurate. Instead of
        -- In this case, we have already checked a valid QC for the round and epoch of the timeout
        -- message, but the message is on either a 'BlockPending' or 'BlockUnknown' block.
        -- We just need to process the timeout.

We could write
        -- In this case, we have already checked a valid QC for the round and epoch of the timeout
        -- message. We just need to process the timeout.

File: concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs [hunk 1]
```diff
@@ -143,41 +143,38 @@ receiveTimeoutMessage tm@TimeoutMessage{tmBody = TimeoutMessageBody{..}} skovDat
...
+                        -- our highest certified block.
+                        let existingQC = skovData ^. roundExistingQuorumCertificate (qcRound tmQuorumCertificate)
+                            oldOrCurrentHighestCertifiedQCRound = qcRound tmQuorumCertificate <= skovData ^. roundStatus . rsHighestCertifiedBlock . to cbRound
+                        case (existingQC, oldOrCurrentHighestCertifiedQCRound) of
+                            (Just (QuorumCertificateCheckedWitness certEpoch), True) -> do
+                                -- The epochs must match otherwise we reject.
+                                if qcEpoch tmQuorumCertificate == certEpoch
+                                    then received finInfo finalizationCommittee Absent
+                                    else return $ Rejected BadQCEpoch
+                            -- If we have not already checked a QC for the round of the quorum certificate assoicated
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs:158]:
```suggestion
                            -- If we have not already checked a QC for the round of the quorum certificate associated
```

File: concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs [hunk 2]
```diff
@@ -143,41 +143,38 @@ receiveTimeoutMessage tm@TimeoutMessage{tmBody = TimeoutMessageBody{..}} skovDat
...
+                            -- with the timeout message.
+                            _ -> do
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs:160]: Does it ever make sense to log the distinction between the 3 different scenarios here?

File: concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs [hunk 3]
```diff
@@ -319,7 +315,7 @@ executeTimeoutMessage (PartiallyVerifiedTimeoutMessage{..})
                 then -- Advance the round with the new certified block.
                     advanceRoundWithQuorum newCertifiedBlock
                 else -- Otherwise we just update the highest certified block.
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs:317]: I suppose the comment isn't exactly accurate anymore.

**Review by abizjak — APPROVED** [review]

File: concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs (hunk 3 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs:317]: Fixed.

**Pull request merged**